### PR TITLE
refactor(middleware): make middleware settings typesafe

### PR DESCRIPTION
**Pull request by gins3000**
On my quest to introduce a "piped dispatch", I got side tracked, and saw that middleware settings were not typesafe, so I made them so.

Before this change
```ts
function myMiddleware(newState: State, oldState?: State, settings?: { foo: string }) { ... }

store.registerMiddleware(myMiddleware, MiddlewarePlacement.After, { notFoo: 42 }); // bad!
```
would have been accepted.
But with this change, the build would fail, and make you fix the code to one of these:
```ts
store.registerMiddleware(myMiddleware, MiddlewarePlacement.After, { foo: "some string" });
store.registerMiddleware(myMiddleware, MiddlewarePlacement.After);
```
Bonus: as a middleware writer, you can now force settings to be passed, by just making them a non-optional parameter in your middleware:
```ts
function forcedSettingsMiddleware(newState: State, oldState: State | undefined, settings: { foo: string }) { ... }

store.registerMiddleware(forcedSettingsMiddleware, MiddlewarePlacement.After); // would now scream
store.registerMiddleware(forcedSettingsMiddleware, MiddlewarePlacement.After, { foo: "some string" }); // would work
```

Double bonus: ff you have at least a certain TypeScript version which fixed some inference behavior (I believe 3.3.0+?), you even can Ctrl+Space inside the settings to get auto-completion. But for just finding type mismatches, the current minimum TypeScript version 3.0.1 works just as well.
In short, this change should be **non-breaking**, but enhanced if a higher TypeScript version is used.
 
I hope you find this useful. If there is anything that needs to be improved or changed, please let me know :)

For implementation details, please see the commit messages.

PS: I tried to sign the CLA but the link in https://github.com/aurelia/store/blob/master/CONTRIBUTING.md points to a dead googledoc.

### CODE REVIEW

**Comment by zewa666** [comment]
Thx for the contribution
